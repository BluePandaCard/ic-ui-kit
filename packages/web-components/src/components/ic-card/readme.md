# ic-card



<!-- Auto Generated Below -->


## Properties

| Property               | Attribute        | Description                                                                                     | Type                                                                                                                                                                                     | Default     |
| ---------------------- | ---------------- | ----------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `clickable`            | `clickable`      | Set the variant to static or clickable                                                          | `boolean`                                                                                                                                                                                | `false`     |
| `disabled`             | `disabled`       | Whether the card is disabled if it is clickable                                                 | `boolean`                                                                                                                                                                                | `false`     |
| `fullWidth`            | `full-width`     | If true then fill width of container                                                            | `boolean`                                                                                                                                                                                | `false`     |
| `heading` _(required)_ | `heading`        | Title for the card                                                                              | `string`                                                                                                                                                                                 | `undefined` |
| `href`                 | `href`           | URL for clickable cards                                                                         | `string`                                                                                                                                                                                 | `undefined` |
| `hreflang`             | `hreflang`       | Human readable explanation of the URL                                                           | `string`                                                                                                                                                                                 | `""`        |
| `message`              | `message`        | Main body message of the card                                                                   | `string`                                                                                                                                                                                 | `""`        |
| `referrerpolicy`       | `referrerpolicy` | How much of the referrer to send when following the link.                                       | `"" \| "no-referrer" \| "no-referrer-when-downgrade" \| "origin" \| "origin-when-cross-origin" \| "same-origin" \| "strict-origin" \| "strict-origin-when-cross-origin" \| "unsafe-url"` | `undefined` |
| `rel`                  | `rel`            | The relationship of the linked URL as space-separated link types.                               | `string`                                                                                                                                                                                 | `undefined` |
| `target`               | `target`         | Where to display the linked URL, as the name for a browsing context (a tab, window, or iframe). | `string`                                                                                                                                                                                 | `undefined` |


## Slots

| Slot        | Description                                                             |
| ----------- | ----------------------------------------------------------------------- |
| `"heading"` | Content will be placed at the top of the card to the right of the icon. |
| `"icon"`    | Content will be placed to the left of the card title.                   |
| `"message"` | Content will be placed in the main body of the card.                    |


## Dependencies

### Depends on

- [ic-typography](../ic-typography)

### Graph
```mermaid
graph TD;
  ic-card --> ic-typography
  style ic-card fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


