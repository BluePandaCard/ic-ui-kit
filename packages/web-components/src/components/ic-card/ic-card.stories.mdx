import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Description,
} from "@storybook/addon-docs";
import { html } from "lit-html";
import readme from "./readme.md";
import StatusTag from "../ic-status-tag/readme.md";
import Typography from "../ic-typography/readme.md";

<Meta title="Web Components/Card" component="ic-card" />

<Description markdown={readme} />

### Static

<Canvas>
  <Story name="Static">
    {html`<ic-card
      heading="Card"
      message="This is an example of a new card component with text included"
      ><svg
        slot="icon"
        xmlns="http://www.w3.org/2000/svg"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        fill="#000000"
      >
        <path d="M0 0h24v24H0V0z" fill="none" />
        <path
          d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
        /></svg
    ></ic-card>`}
  </Story>
</Canvas>

### Title only

<Canvas>
  <Story name="Title only">
    {html`<ic-card heading="Joe BLOGGS 123" clickable="true"
      ><svg
        slot="icon"
        xmlns="http://www.w3.org/2000/svg"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        fill="#000000"
      >
        <path d="M0 0h24v24H0V0z" fill="none" />
        <path
          d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
        /></svg
    ></ic-card>`}
  </Story>
</Canvas>

### Clickable as link

<Canvas>
  <Story name="Clickable as link">
    {html`<ic-card
      heading="Card"
      message="This is an example of a new card component with text included"
      clickable="true"
      href="https://www.google.com"
      ><svg
        slot="icon"
        xmlns="http://www.w3.org/2000/svg"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        fill="#000000"
      >
        <path d="M0 0h24v24H0V0z" fill="none" />
        <path
          d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
        /></svg
    ></ic-card>`}
  </Story>
</Canvas>

### Clickable as button

<Canvas>
  <Story name="Clickable as button">
    {html`<ic-card
      heading="Card"
      message="This is an example of a new card component with text included"
      clickable="true"
      onclick="alert('clicked')"
      ><svg
        slot="icon"
        xmlns="http://www.w3.org/2000/svg"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        fill="#000000"
      >
        <path d="M0 0h24v24H0V0z" fill="none" />
        <path
          d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
        /></svg
    ></ic-card>`}
  </Story>
</Canvas>

### Disabled

<Canvas>
  <Story name="Disabled">
    {html`<ic-card
      heading="Card"
      message="This is an example of a new card component with text included"
      clickable="true"
      disabled="true"
      onClick="alert('clicked')"
      ><svg
        slot="icon"
        xmlns="http://www.w3.org/2000/svg"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        fill="#000000"
      >
        <path d="M0 0h24v24H0V0z" fill="none" />
        <path
          d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
        /></svg
    ></ic-card>`}
  </Story>
</Canvas>

### Using slots

<Canvas>
  <Story name="Using slots">
    {html` <ic-card
        message="This card uses a slot to display a status tag in place of the heading"
      >
        <ic-status-tag
          slot="heading"
          status="success"
          label="Active"
        ></ic-status-tag>
      </ic-card>
      <ic-card heading="Card">
        <div slot="message" style="display:flex;align-items:center;">
          <ic-typography variant="p" style="margin-right:8px;"
            >This card uses a slot to display its message with a status
            tag</ic-typography
          >
          <ic-status-tag status="success" label="Active"></ic-status-tag>
        </div>
      </ic-card>`}
  </Story>
</Canvas>

### Full width

<Canvas>
  <Story name="Full width">
    {html`
      <ic-card
        heading="Card"
        message="This is an example of a card component that fills it's parent container"
        clickable="true"
        full-width="true"
        href="https://www.google.com"
        ><svg
          slot="icon"
          xmlns="http://www.w3.org/2000/svg"
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
          fill="#000000"
        >
          <path d="M0 0h24v24H0V0z" fill="none" />
          <path
            d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
          /></svg
      ></ic-card>
    `}
  </Story>
</Canvas>

### Wrapped by link

<Canvas>
  <Story name="Wrapped by link">
    {html`<a href="/"
      ><ic-card
        heading="Card"
        message="This is an example of a new card component with text included"
        ><svg
          slot="icon"
          xmlns="http://www.w3.org/2000/svg"
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
          fill="#000000"
        >
          <path d="M0 0h24v24H0V0z" fill="none" />
          <path
            d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
          /></svg></ic-card
    ></a>`}
  </Story>
</Canvas>
