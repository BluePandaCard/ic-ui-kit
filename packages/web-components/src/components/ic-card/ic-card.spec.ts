import { newSpecPage } from "@stencil/core/testing";
import { Card } from "./ic-card";

describe("ic-card", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<ic-card heading="Card" message="This is a static card"></ic-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render as a button", async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<ic-card heading="Card" message="This is a clickable card rendered as a button" clickable=true></ic-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render as a link", async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<ic-card heading="Card" message="This is a clickable card rendered as a button" clickable=true href="/"></ic-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render full width variant", async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<ic-card heading="Card" full-width=true message="This is a full width card" clickable=true></ic-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with a link parent", async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<a href="/"><ic-card heading="Card" message="This is a clickable card" clickable=true></ic-card></a>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should apply 'focussed' style when parent is focussed", async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<a href="/"><ic-card id="test-id" heading="Card" message="This is a clickable card" clickable=true></ic-card></a>`,
    });

    expect(page.root).not.toBeNull;

    const element = await document.getElementById("test-id");

    await expect(element.classList.contains("focussed")).toBeFalsy;

    await element.focus();

    await expect(element.classList.contains("focussed")).toBeTruthy;
  });

  it("should lose 'focussed' style when parent loses focus", async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<a href="/"><ic-card id="test-card" heading="Card" message="This is a clickable card" clickable=true></ic-card></a>`,
    });

    expect(page.root).not.toBeNull;

    const element = await document.getElementById("test-card");

    await element.focus();

    await expect(element.classList.contains("focussed")).toBeTruthy;

    await element.blur();

    await expect(element.classList.contains("focussed")).toBeFalsy;

    await page.rootInstance.disconnectedCallback();
  });

  it("should stop immediate propagation of a click event when disabled", async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<ic-card id="test-card" heading="Card" message="This is a clickable card" clickable=true disabled=true onclick="alert('test')"></ic-card>`,
    });

    jest.spyOn(window, "alert").mockImplementation();

    const element = await document.getElementById("test-card");

    element.click();

    await page.waitForChanges();

    await expect(window.alert).not.toHaveBeenCalled;
  });
});
