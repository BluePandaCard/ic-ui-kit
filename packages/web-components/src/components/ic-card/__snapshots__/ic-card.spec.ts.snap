// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-card should render 1`] = `
<ic-card heading="Card" message="This is a static card">
  <mock:shadow-root>
    <div class="card">
      <div class="card-header">
        <div class="card-title">
          <slot name="heading">
            <ic-typography variant="h4">
              <p>
                Card
              </p>
            </ic-typography>
          </slot>
        </div>
      </div>
      <div class="card-message">
        <slot name="message">
          <ic-typography variant="body">
            This is a static card
          </ic-typography>
        </slot>
      </div>
    </div>
  </mock:shadow-root>
</ic-card>
`;

exports[`ic-card should render as a button 1`] = `
<ic-card clickable="true" heading="Card" message="This is a clickable card rendered as a button">
  <mock:shadow-root>
    <button class="card clickable" tabindex="0">
      <div class="card-header">
        <div class="card-title">
          <slot name="heading">
            <ic-typography variant="h4">
              <p>
                Card
              </p>
            </ic-typography>
          </slot>
        </div>
      </div>
      <div class="card-message">
        <slot name="message">
          <ic-typography variant="body">
            This is a clickable card rendered as a button
          </ic-typography>
        </slot>
      </div>
    </button>
  </mock:shadow-root>
</ic-card>
`;

exports[`ic-card should render as a link 1`] = `
<ic-card clickable="true" heading="Card" href="/" message="This is a clickable card rendered as a button">
  <mock:shadow-root>
    <a class="card clickable" href="/" hreflang="" tabindex="0">
      <div class="card-header">
        <div class="card-title">
          <slot name="heading">
            <ic-typography variant="h4">
              <p>
                Card
              </p>
            </ic-typography>
          </slot>
        </div>
      </div>
      <div class="card-message">
        <slot name="message">
          <ic-typography variant="body">
            This is a clickable card rendered as a button
          </ic-typography>
        </slot>
      </div>
    </a>
  </mock:shadow-root>
</ic-card>
`;

exports[`ic-card should render full width variant 1`] = `
<ic-card clickable="true" full-width="true" heading="Card" message="This is a full width card">
  <mock:shadow-root>
    <button class="card clickable fullwidth" tabindex="0">
      <div class="card-header">
        <div class="card-title">
          <slot name="heading">
            <ic-typography variant="h4">
              <p>
                Card
              </p>
            </ic-typography>
          </slot>
        </div>
      </div>
      <div class="card-message">
        <slot name="message">
          <ic-typography variant="body">
            This is a full width card
          </ic-typography>
        </slot>
      </div>
    </button>
  </mock:shadow-root>
</ic-card>
`;

exports[`ic-card should render with a link parent 1`] = `
<ic-card clickable="true" heading="Card" message="This is a clickable card">
  <mock:shadow-root>
    <div class="card clickable">
      <div class="card-header">
        <div class="card-title">
          <slot name="heading">
            <ic-typography variant="h4">
              <p>
                Card
              </p>
            </ic-typography>
          </slot>
        </div>
      </div>
      <div class="card-message">
        <slot name="message">
          <ic-typography variant="body">
            This is a clickable card
          </ic-typography>
        </slot>
      </div>
    </div>
  </mock:shadow-root>
</ic-card>
`;
